pallets/author-mapping/src/lib.rs

```diff
@@ -125,167 +144,157 @@ pub mod pallet {
 			let account_id = ensure_signed(origin)?;
+			ensure!(keys.len() == keys_size::<T>(), Error::<T>::WrongKeySize);
+			let encoded = &mut keys.as_slice();
+			let new_nimbus_id =
+				NimbusId::decode(encoded).map_err(|_| Error::<T>::DecodeNimbusFailed)?;
+			let keys = T::Keys::decode(encoded).map_err(|_| Error::<T>::DecodeKeysFailed)?;
+
+			if let Some(old_nimbus_id) = Self::nimbus_id_of(&account_id) {
+				Self::rotate_keys(old_nimbus_id, new_nimbus_id, account_id, keys)
+			} else {
+				Self::register_keys(new_nimbus_id, account_id, keys)
+			}
+		}
+	}
 
-			let stored_info = MappingWithDeposit::<T>::try_get(&author_id)
+	impl<T: Config> Pallet<T> {
+		fn rm_keys(nimbus_id: NimbusId, account_id: T::AccountId) -> DispatchResult {
+			let stored_info = MappingWithDeposit::<T>::try_get(&nimbus_id)
 				.map_err(|_| Error::<T>::AssociationNotFound)?;
 
 			ensure!(
 				account_id == stored_info.account,
 				Error::<T>::NotYourAssociation
 			);
 
-			MappingWithDeposit::<T>::remove(&author_id);
+			MappingWithDeposit::<T>::remove(&nimbus_id);
 			NimbusLookup::<T>::remove(&account_id);
 
 			T::DepositCurrency::unreserve(&account_id, stored_info.deposit);
 
-			<Pallet<T>>::deposit_event(Event::AuthorDeRegistered {
-				author_id,
+			<Pallet<T>>::deposit_event(Event::KeysRemoved {
+				nimbus_id,
 				account_id,
 				keys: stored_info.keys,
 			});
-
-			Ok(().into())
+			Ok(())
 		}
-
-		/// Add association and set session keys
-		#[pallet::weight(<T as Config>::WeightInfo::register_keys())]
-		pub fn register_keys(origin: OriginFor<T>, keys: (NimbusId, T::Keys)) -> DispatchResult {
-			let account_id = ensure_signed(origin)?;
-			let (author_id, keys) = keys;
+		fn register_keys(
+			nimbus_id: NimbusId,
+			account_id: T::AccountId,
+			keys: T::Keys,
+		) -> DispatchResult {
 			ensure!(
-				MappingWithDeposit::<T>::get(&author_id).is_none(),
+				MappingWithDeposit::<T>::get(&nimbus_id).is_none(),
 				Error::<T>::AlreadyAssociated
 			);
+			Self::enact_registration(&nimbus_id, &account_id, keys.clone())?;
 
-			Self::enact_registration(&author_id, &account_id, keys.clone())?;
-
-			<Pallet<T>>::deposit_event(Event::AuthorRegistered {
-				author_id,
+			<Pallet<T>>::deposit_event(Event::KeysRegistered {
+				nimbus_id,
 				account_id,
 				keys,
 			});
-
 			Ok(())
 		}
-
-		/// Set association and session keys at once.
-		///
-		/// This is useful for key rotation to update Nimbus and VRF keys in one call.
-		/// No new security deposit is required. Will replace `update_association` which is kept
-		/// now for backwards compatibility reasons.
-		#[pallet::weight(<T as Config>::WeightInfo::set_keys())]
-		pub fn set_keys(origin: OriginFor<T>, keys: (NimbusId, T::Keys)) -> DispatchResult {
-			let account_id = ensure_signed(origin)?;
-			let old_author_id =
-				Self::nimbus_id_of(&account_id).ok_or(Error::<T>::OldAuthorIdNotFound)?;
-			let stored_info = MappingWithDeposit::<T>::try_get(&old_author_id)
+		fn rotate_keys(
```

Comment:
Now that this became a helper fn, it lost the ability to internally ensure that the origin is properly signed by the holder of the account. It's now the caller's responsibility to do this.

The only way I see to "fix" that is to forward the origin on from the caller to check it here. If that's undesireable, this should at least be documented with the fn.

Reply:
I've reviewed both cases of calling `rotate_keys`, they both properly check origin :)